### DISCUSSION TITLE
Supported python versions

### DISCUSSION THREAD

**Discussion by robbievanleeuwen** [robbievanleeuwen/section-properties#82] [Ideas]
Hi all,

Off the back of a previous discussion #61, I am considering removing support for python 3.6 to allow supporting the recent updates to scipy (1.6 & 1.7). 

Unfortunately I don't think providing support for python 3.8 is possible at this time as there are no wheels for Windows users available at [https://www.lfd.uci.edu/~gohlke/pythonlibs/#meshpy](https://www.lfd.uci.edu/~gohlke/pythonlibs/#meshpy)... Windows strikes again :facepalm:

Hopefully I can get rid of meshpy and implement [triangle](https://github.com/drufat/triangle) once v2 drops and we can get a bit more version flexibility!

Let me know if you have any thoughts on only supporting python 3.7!

Cheers,
Robbie

**Comment by Czarified** [comment]
Hi Robbie! Good to hear from you again!

I'm using Python 3.8.5 and I don't see any issues with sectionproperties or MeshPy. On my installed version:
```
(base) C:\Users\bcrews\Documents\PythonScripts\boa_testing>pytest --pyargs sectionproperties
======================================================= test session starts =======================================================
platform win32 -- Python 3.8.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: C:\Users\bcrews\Documents\PythonScripts\boa_testing
plugins: anyio-2.2.0, xonsh-0.9.27
collected 5 items

tests\test_rectangle.py ...                                                                                                  [ 60%]
tests\test_validation.py ..                                                                                                  [100%]

======================================================== warnings summary =========================================================
..\..\..\.venv\base\lib\site-packages\scipy\special\orthogonal.py:79
..\..\..\.venv\base\lib\site-packages\scipy\special\orthogonal.py:79
  c:\users\bcrews\.venv\base\lib\site-packages\scipy\special\orthogonal.py:79: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    from numpy import (exp, inf, pi, sqrt, floor, sin, cos, around, int,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================= 5 passed, 2 warnings in 36.75s ==================================================
```

I assume upgrading scipy will get rid of those warnings. I also have a personal library with more meshing and testing (almost exactly the same as the tests in #54 ), and it passes with the exact same warnings.

**Reply by robbievanleeuwen** [reply]
Hi @Czarified,

May I ask how you installed meshpy? A couple of years ago it wasn't possible to install meshpy through pip if you were using Windows.  This is why I recommended windows users perform the meshpy wheel installation separate to `pip install sectionproperties`. If it's working now maybe we can incorporate 3.8 as well!

Robbie

**Reply by Czarified** [reply]
My procedure was: Install [Microsoft Visual Studio Build Tools](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2019) (the C++ compiler), then pip install meshpy, then pip install sectionproperties. I have used a requirements.txt to repeatedly setup environments, and just alphabetically meshpy is always first.

**Reply by robbievanleeuwen** [reply]
Sorry for my ignorance as I don't have a windows OS - would the above procedure fail if you didn't have the c++ compiler first installed?

My goal at the moment would be to include meshpy in the windows setup requirements so a new user could simply run `pip install sectionproperties` just like a mac/linux user without the need to perform additional steps. My concern is that a new user (with a fresh windows installation) would run into problems installing _sectionproperties_ because they don't have a c++ compiler installed.

The work around at the moment is to use the meshpy wheels which seem to only be supported up to python 3.7.

I guess at the end of the day I want to try to make the installation as simple for windows users as possible and try to find a 'one size fits all' solution.

**Reply by Czarified** [reply]
If you try and install meshpy on Windows without the c++ compiler, pip will error out and tell you to install Visual Studio, with a link to Microsoft's download page. You just don't actually need _all_ of visual studio, only the compiler, which is freely available and (from what I can tell) "openly" licensed. I'll see if I can spin up a virtual machine and get the complete error message...

I'm onboarding a bunch of analysts at work, and we intend on using sectionproperties. So part of our install process (since we're all on Windows) is getting the VS Build Tools installed in the background. It's a bit of a pain, and introduces more places people can mess up, but it hasn't caused too many problems so far.

**Reply by robbievanleeuwen** [reply]
Ok thanks for this, that's a good installation suggestion I could put in the docs for those struggling with windows!

**Reply by connorferster** [reply]
@Czarified Ah! That must be why my meshpy installed so easily: I had Visual Studio Build Tools installed.

**Comment by Agent6-6-6** [comment]
@robbievanleeuwen , also on Windows Python 3.8.10 here with meshpy (for quite a while it has worked, installed via pip).
`pip install MeshPy`
`pip install --upgrade MeshPy`

Something doesn't work with Python 3.9.x, Scipy I think.

**Reply by robbievanleeuwen** [reply]
Ok thanks for the heads up! Will look to implementing both 3.7 & 3.8, while dropping 3.6!

**Comment by Spectre5** [comment]
I think this will be a non-issue when/if we switch to using the [triangle](https://github.com/drufat/triangle) package, which appears to have [wheels](https://pypi.org/project/triangle/#files) for all 3 OSes on PyPI.  It does not appear to be really active, but it may not need many changes so that could be ok, so long as it publishes new wheel files as new Python versions come out in the future.  If there are issues with this package, then I think the best choice would be to build the wheels ourselves for each OS using GitHub actions and [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/), which is what triangle [uses](https://github.com/drufat/triangle/blob/master/.github/workflows/wheels.yml) as well.  It wouldn't take much to convert the code to use triangle, see their documents for some examples including ones with holes.

Otherwise, we just require users to have the free compiler on Windows installed, which will also work fine.

**Reply by robbievanleeuwen** [reply]
Agreed, was wondering if it was worth doing an interim version bump in between now and when triangle is implemented (post v2). I was thinking I could release a 1.0.9 in the next few days with a few of the bug fixes and minor new features over the last 9 months, as well as now supporting 3.7 & 3.8, while dropping 3.6. This could transition nicely into the imminent release of v2. I think your last line is a good interim measure for windows users.

**Reply by robbievanleeuwen** [reply]
Think I'm going to scrap the 1.0.9 idea and go all in with v2.

**Comment by DougAJ4** [comment]
Hi Robbie, after a bit of a break I have been working on linking sectionproperties to Excel using pyxll (following a recent thread at Eng-tips).

I am using Python 3.9.7 on Windows.  I had no problem installing meshpy, but when I tried to install sectionproperties using pip it looked for progressively earlier versions of numpy, and I eventually cancelled that process.  However I found that if I just copied the sectionproperties folder into my active pyxll development folder everything worked without a problem.

I'll post more details on my blog in a few days.

**Reply by robbievanleeuwen** [reply]
Hi @DougAJ4, thanks for the info. *sectionproperties* v2 (to be released in the next couple of days) will support python 3.7, 3.8 & 3.9. *meshpy* is being dropped as a dependency in favour of [triangle](https://github.com/drufat/triangle) which should pose significantly less installation issues, particularly for windows users but essentially performs exactly the same task as *meshpy* (a wrapper for Triangle).

**Reply by DougAJ4** [reply]
Just installed v2 (+ triangle etc) with pip with no problems at all.
Now updating my Excel code, which is all working so far.

**Reply by robbievanleeuwen** [reply]
Glad to hear @DougAJ4! Looks like this issue with triangle & python 3.9 has been identified - see #172.
